File: server/internal/game/building_defs.go
package game

import (
	"embed"
	"encoding/json"
	"io/fs"
	"sort"
	"strings"
)

//go:embed all:catalog/buildings
var buildingDefsFS embed.FS

// BuildingDef holds the configuration for a buildable building type.
// Client-only fields (Color, Label, Hotkey) are passed through to the
// API as-is; the server game logic never reads them. Procedural render
// fallbacks are no longer carried in the catalog — they live frontend-side
// (see client buildingFallbackRender.ts).
type BuildingDef struct {
	Type         string  `json:"type"`
	Class        string  `json:"class,omitempty"`
	Buildable    *bool   `json:"buildable,omitempty"`
	Width        int     `json:"width"`
	Height       int     `json:"height"`
	MaxHp        float64 `json:"maxHp"`
	BuildSeconds float64 `json:"buildSeconds"`
	Damage       int     `json:"damage,omitempty"`
	AttackRange  float64 `json:"attackRange,omitempty"`
	AttackSpeed  float64 `json:"attackSpeed,omitempty"`
	VisionRange  float64 `json:"visionRange,omitempty"`
	// UnobstructedVision, when true, makes this building's FOW vision ignore
	// line-of-sight blockers (trees/obstacles and terrain cliffs) within its
	// range — it sees over them, the way flyer units do. Omitted ⇒ normal,
	// occluded vision. Set on Tower so a tower placed in a forest still reveals
	// its full radius.
	UnobstructedVision bool            `json:"unobstructedVision,omitempty"`
	AttackVisual       json.RawMessage `json:"attackVisual,omitempty"`
	ResourceType       string          `json:"resourceType,omitempty"`
	ResourceAmount     int             `json:"resourceAmount,omitempty"`
	ResourceCost       map[string]int  `json:"resourceCost"`
	Capabilities       []string        `json:"capabilities"`
	SpawnUnitTypes     []string        `json:"spawnUnitTypes"`
	// RequiresTownhallTier gates construction: the owning player must control
	// at least one fully-built townhall whose tier is ≥ this value before
	// BuildBuilding accepts the placement. Zero/omitted ⇒ no requirement.
	// Tiers: 1 = Town Hall, 2 = Keep, 3 = Castle (mirrors the upgrade chain
	// in state_upgrades.go's handleUpgradeBuildingTierLocked).
	RequiresTownhallTier int `json:"requiresTownhallTier,omitempty"`
	// UpgradesFrom names the building type this one is a tier-up of (e.g. keep
	// upgradesFrom townhall, castle upgradesFrom keep). Empty for base/standalone
	// buildings. The defs form an ordered chain that the tier-up logic walks
	// instead of hardcoding tiers. NOTE: a placed building keeps its base type
	// plus a numeric `tier` in metadata — it does not change BuildingType on
	// upgrade. These tier defs supply the per-step cost, duration, and display
	// name; their other stats currently mirror the base (tiers are stat-identical
	// today) and are the hook for future per-tier balancing.
	UpgradesFrom string `json:"upgradesFrom,omitempty"`
	// UpgradeCost is the resource cost to upgrade INTO this tier (paired with
	// UpgradesFrom). Ignored on base buildings.
	UpgradeCost map[string]int `json:"upgradeCost,omitempty"`
	// UpgradeSeconds is how long the tier-up into this building takes.
	UpgradeSeconds float64        `json:"upgradeSeconds,omitempty"`
	Metadata       map[string]any `json:"metadata"`
	Color          string         `json:"color,omitempty"`
	Label          string         `json:"label,omitempty"`
	Hotkey         string         `json:"hotkey,omitempty"`
	// SpriteRender lets a building's sprite extend beyond its grid footprint
	// without affecting pathing, selection hit-testing, or the grid cells
	// the building occupies. Mirrors ObstacleRenderDef semantics (cell units,
	// omitted fields use footprint defaults). Used for e.g. a barracks with
	// a flag pole taller than its footprint, or a wider sprite that spills
	// half a cell over each side.
	SpriteRender *BuildingSpriteRenderDef `json:"spriteRender,omitempty"`
	// Shadow tunes the ground shadow drawn under the building. Client-only
	// render config; the server never reads it and only passes it through.
	// Omit the block to use the footprint-derived default shadow.
	Shadow *BuildingShadowDef `json:"shadow,omitempty"`
	// SelectionRing nudges the size/placement of the selection & hover ring
	// drawn around the building. Client-only (the server never reads it).
	// Mirrors ObstacleSelectionRingDef semantics. Omit the block to keep the
	// default ring derived from the footprint width.
	SelectionRing *BuildingSelectionRingDef `json:"selectionRing,omitempty"`
}

// BuildingSelectionRingDef sizes/places the selection & hover ellipse for a
// building. All fields are in *cell units*. Defaults (when a field is zero /
// omitted) reproduce the legacy footprint-derived ring:
//
//	radiusX = footprintWidth * 0.55
//	radiusY = radiusX * 0.34
//	center  = horizontally centered, ~0.62 down the footprint
//
// offsetX/offsetY (when set) are measured from the footprint's top-left
// corner. Setting only radiusX/radiusY enlarges the ring while keeping the
// default centering.
type BuildingSelectionRingDef struct {
	OffsetX float64 `json:"offsetX,omitempty"`
	OffsetY float64 `json:"offsetY,omitempty"`
	RadiusX float64 `json:"radiusX,omitempty"`
	RadiusY float64 `json:"radiusY,omitempty"`
}

// BuildingSpriteRenderDef is the sprite-overflow config for a building.
// Shape parallels ObstacleRenderDef. All fields are in cell units; zero
// means "use default" (offset=0, width/height = footprint).
type BuildingSpriteRenderDef struct {
	OffsetX float64 `json:"offsetX,omitempty"`
	OffsetY float64 `json:"offsetY,omitempty"`
	Width   float64 `json:"width,omitempty"`
	Height  float64 `json:"height,omitempty"`
}

// BuildingShadowDef tunes the ground shadow drawn under a building. All
// distance fields are in *cell units*; omitted/zero fields fall back to a
// footprint-derived default (see resolveBuildingShadow on the client):
//
//	radiusX = footprintWidth * 0.42
//	radiusY = radiusX * 0.3
//	center  = horizontally centered, near the footprint's bottom edge
//
// Client-only render config; the server passes it through untouched.
type BuildingShadowDef struct {
	// Enabled defaults to true; set false to draw no shadow for this building.
	Enabled *bool   `json:"enabled,omitempty"`
	OffsetX float64 `json:"offsetX,omitempty"`
	OffsetY float64 `json:"offsetY,omitempty"`
	RadiusX float64 `json:"radiusX,omitempty"`
	RadiusY float64 `json:"radiusY,omitempty"`
	// Opacity is the peak alpha at the shadow center (0..1). Zero/omitted ⇒
	// the client default.
	Opacity float64 `json:"opacity,omitempty"`
}

// BuildingStyleRenderDef is a per-art render override for a building that
// selects its sprite per-instance rather than from a single sprite.png
// (currently only recipe-shop, via the shopStyle metadata set in the map
// editor). It carries only the client-render fields that differ between art
// variants — the gameplay def in <type>/<type>.json remains the single source
// of truth for footprint, capabilities, HP, etc. Authored as one JSON per
// sprite, colocated with the base def and keyed by the file stem (= style
// name = the sprite's file stem under client assets/buildings/recipe-shops/).
// The server never reads these fields; it passes them through to the client,
// which prefers them over the base def's render config when a style is set.
type BuildingStyleRenderDef struct {
	SpriteRender  *BuildingSpriteRenderDef  `json:"spriteRender,omitempty"`
	SelectionRing *BuildingSelectionRingDef `json:"selectionRing,omitempty"`
	Shadow        *BuildingShadowDef        `json:"shadow,omitempty"`
}

// Class returns the building's class, defaulting to "player" when unspecified.
func (d BuildingDef) ClassOrDefault() string {
	if d.Class == "" {
		return "player"
	}
	return d.Class
}

// HpPerSecond returns the build/repair rate for this building type.
func (d BuildingDef) HpPerSecond() float64 {
	if d.BuildSeconds <= 0 {
		return d.MaxHp
	}
	return d.MaxHp / d.BuildSeconds
}

// buildingCatalog / buildingDefsByType MUST be var initializers (not init())
// so that Go's package-level dependency analysis can order them before any
// other var initializer that calls getBuildingDef — most importantly
// unitDefsByType's loader, which validates each unit's RequiresBuildings
// against the building catalog. All package-level var initializers run
// before any init() function, so converting this to init() would race
// with unitDefsByType (and break maps.go's placedUnits hydration via
// the `_ = unitDefsByType` dependency-injection marker). Go orders these by
// reference: unitDefsByType → getBuildingDef → buildingDefsByType →
// buildingCatalog.
var buildingCatalog = loadBuildingCatalog()
var buildingDefsByType = buildingCatalog.defs
var buildingStyleRenders = buildingCatalog.styles

type buildingCatalogData struct {
	defs map[string]BuildingDef
	// styles maps buildingType → styleName → per-style render override, for
	// buildings authored as a subdirectory (see loadBuildingCatalog).
	styles map[string]map[string]BuildingStyleRenderDef
}

// loadBuildingCatalog reads catalog/buildings. Two authoring layouts coexist:
//
//   - A flat <type>.json at the top level is a single BuildingDef (the common
//     case — one building type, one shared sprite).
//   - A subdirectory <type>/ groups a building that picks its sprite
//     per-instance: <type>/<type>.json is the gameplay BuildingDef, and every
//     other <type>/<style>.json is a BuildingStyleRenderDef keyed by its file
//     stem (the style name). This lets each sprite variant carry its own
//     selection ring / sprite bounds while sharing one gameplay def.
func loadBuildingCatalog() buildingCatalogData {
	const root = "catalog/buildings"
	entries, err := fs.ReadDir(buildingDefsFS, root)
	if err != nil {
		panic(root + ": " + err.Error())
	}
	defs := make(map[string]BuildingDef, len(entries))
	styles := make(map[string]map[string]BuildingStyleRenderDef)

	loadDef := func(path string) {
		data, err := buildingDefsFS.ReadFile(path)
		if err != nil {
			panic(path + ": " + err.Error())
		}
		var def BuildingDef
		if err := json.Unmarshal(data, &def); err != nil {
			panic(path + ": " + err.Error())
		}
		defs[def.Type] = def
	}

	for _, entry := range entries {
		name := entry.Name()
		if !entry.IsDir() {
			if strings.HasSuffix(name, ".json") {
				loadDef(root + "/" + name)
			}
			continue
		}
		dir := root + "/" + name
		subEntries, err := fs.ReadDir(buildingDefsFS, dir)
		if err != nil {
			panic(dir + ": " + err.Error())
		}
		baseFile := name + ".json"
		for _, sub := range subEntries {
			if sub.IsDir() || !strings.HasSuffix(sub.Name(), ".json") {
				continue
			}
			path := dir + "/" + sub.Name()
			if sub.Name() == baseFile {
				loadDef(path)
				continue
			}
			data, err := buildingDefsFS.ReadFile(path)
			if err != nil {
				panic(path + ": " + err.Error())
			}
			var style BuildingStyleRenderDef
			if err := json.Unmarshal(data, &style); err != nil {
				panic(path + ": " + err.Error())
			}
			if styles[name] == nil {
				styles[name] = make(map[string]BuildingStyleRenderDef)
			}
			styles[name][strings.TrimSuffix(sub.Name(), ".json")] = style
		}
	}
	return buildingCatalogData{defs: defs, styles: styles}
}

func getBuildingDef(buildingType string) (BuildingDef, bool) {
	def, ok := buildingDefsByType[buildingType]
	return def, ok
}

// upgradeChainFor returns the ordered tier chain rooted at rootType: the base
// def first, then each def whose UpgradesFrom points at the previous link.
// e.g. upgradeChainFor("townhall") → [townhall, keep, castle]. Returns nil if
// rootType is unknown. The catalog is authored as a straight chain; the scan is
// deterministic (sorted by type) and capped at the catalog size so a malformed
// cycle can't loop forever.
func upgradeChainFor(rootType string) []BuildingDef {
	root, ok := buildingDefsByType[rootType]
	if !ok {
		return nil
	}

	types := make([]string, 0, len(buildingDefsByType))
	for t := range buildingDefsByType {
		types = append(types, t)
	}
	sort.Strings(types)

	chain := []BuildingDef{root}
	for len(chain) <= len(buildingDefsByType) {
		current := chain[len(chain)-1]
		var next BuildingDef
		found := false
		for _, t := range types {
			if d := buildingDefsByType[t]; d.UpgradesFrom == current.Type {
				next, found = d, true
				break
			}
		}
		if !found {
			break
		}
		chain = append(chain, next)
	}
	return chain
}

// buildingRequirementTier resolves a unit RequiresBuildings entry to the root
// building type and the minimum tier a placed building of that root must reach
// to satisfy it. A tier link (temple, keep, castle) resolves to its chain root
// and 1-based position — "temple" → ("chapel", 2), "keep" → ("townhall", 2). A
// plain type (blacksmith, chapel) or an unknown type resolves to (itself, 1),
// preserving the exact pre-tier behaviour. This exists because a placed
// building keeps its ROOT BuildingType and records progress only in
// metadata["tier"]: a "temple" requirement is really "a chapel at tier ≥ 2".
func buildingRequirementTier(requiredType string) (rootType string, tier int) {
	def, ok := buildingDefsByType[requiredType]
	if !ok {
		return requiredType, 1
	}
	// Walk upgradesFrom back to the chain root (bounded against a malformed cycle).
	root := def
	for i := 0; root.UpgradesFrom != "" && i < len(buildingDefsByType); i++ {
		parent, ok := buildingDefsByType[root.UpgradesFrom]
		if !ok {
			break
		}
		root = parent
	}
	chain := upgradeChainFor(root.Type)
	for i, d := range chain {
		if d.Type == requiredType {
			return root.Type, i + 1
		}
	}
	return requiredType, 1
}

func (d BuildingDef) IsBuildable() bool {
	return d.Buildable == nil || *d.Buildable
}

func ListBuildingDefs() []BuildingDef {
	defs := make([]BuildingDef, 0, len(buildingDefsByType))
	for _, def := range buildingDefsByType {
		defs = append(defs, def)
	}
	sort.Slice(defs, func(i, j int) bool { return defs[i].Type < defs[j].Type })
	return defs
}

// ListBuildingStyleRenders returns the per-style render overrides keyed by
// buildingType → styleName. Consumed by the /catalog/buildings HTTP endpoint
// and passed through to the client renderer, which prefers a style's override
// over the base building def's render config when the instance has that style
// set. Buildings without a subdirectory layout have no entry.
func ListBuildingStyleRenders() map[string]map[string]BuildingStyleRenderDef {
	return buildingStyleRenders
}

File: client/src/game-portal/src/game/maps/buildingDefs.ts
import type { BuildingCapability, JsonObject } from '../network/protocol'

// A rectangular accent layer.
// Coordinates are in cell units relative to the building's top-left origin.
// color: 'player' means "substitute owner/player color at render time".
// kind is optional for backwards compatibility with legacy defs that omit it.
export type BuildingRectLayer = {
  kind?: 'rect'
  x: number
  y: number
  w: number
  h: number
  color: string
}

// A triangle from the 6×6 sub-cell grid.
// cx/cy: cell index within the building footprint.
// sc/sr: sub-column/sub-row within that cell (0–5).
// h: 0 = upper-left triangle (above the / diagonal), 1 = lower-right.
export type BuildingTriLayer = {
  kind: 'tri'
  cx: number
  cy: number
  sc: number
  sr: number
  h: 0 | 1
  color: string
}

export type BuildingRenderLayer = BuildingRectLayer | BuildingTriLayer

// Describes how to visually paint a building.
// inset: padding from the building edge in cell units (default 0.18).
//   Used to anchor HP bars, selection rings, and construction overlays.
// layers: drawn in order, back to front.
export type BuildingRenderDef = {
  inset: number
  layers: BuildingRenderLayer[]
}

// BuildingSpriteRenderDef overrides where and how large a building sprite
// is drawn relative to its grid footprint. Mirrors the obstacle overflow
// system (cell units; omitted fields fall back to the footprint). Grid
// footprint used by pathing, placement, and selection hit-testing is
// unchanged — this only affects the drawn sprite bounding box.
export type BuildingSpriteRenderDef = {
  offsetX?: number
  offsetY?: number
  width?: number
  height?: number
}

// BuildingSelectionRingDef sizes/places the selection & hover ring relative
// to the building's grid footprint. All fields are in cell units. Omitted
// fields fall back to the footprint-derived default ring. Setting just
// radiusX/radiusY enlarges the ring while keeping the default centering.
// Mirrors the Go-side BuildingSelectionRingDef struct.
export type BuildingSelectionRingDef = {
  offsetX?: number
  offsetY?: number
  radiusX?: number
  radiusY?: number
}

// BuildingShadowDef tunes the ground shadow drawn under a building. Distance
// fields are in cell units; omitted/zero fields fall back to a footprint-
// derived default (see resolveBuildingShadow). Mirrors the Go-side struct.
export type BuildingShadowDef = {
  enabled?: boolean
  offsetX?: number
  offsetY?: number
  radiusX?: number
  radiusY?: number
  opacity?: number
}

// Resolved, ready-to-draw building shadow in pixels. center is measured from
// the footprint's top-left world corner.
export type ResolvedBuildingShadow = {
  centerX: number
  centerY: number
  radiusX: number
  radiusY: number
  opacity: number
}

export type BuildingAttackVisual = {
  kind?: 'melee' | 'projectile'
  originX?: number
  originY?: number
  effectLength?: number
}

export type ResolvedBuildingAttackVisual = {
  kind: 'melee' | 'projectile'
  originX: number
  originY: number
  effectLength: number
}

// BuildingStyleRenderDef is a per-art render override for a building that picks
// its sprite per-instance (recipe-shop, via the shopStyle metadata) rather than
// from a single shared sprite. Carries only the render fields that differ
// between art variants; the base BuildingDef stays authoritative for footprint,
// gameplay, etc. Mirrors the Go-side BuildingStyleRenderDef struct. Delivered by
// /catalog/buildings keyed by buildingType → styleName. See getBuildingStyleRender.
export type BuildingStyleRenderDef = {
  spriteRender?: BuildingSpriteRenderDef
  selectionRing?: BuildingSelectionRingDef
  shadow?: BuildingShadowDef
}

export type BuildingClass = 'player' | 'neutral' | 'enemy'

export type BuildingDef = {
  type: string
  class?: BuildingClass
  buildable?: boolean
  width: number
  height: number
  maxHp: number
  buildSeconds: number
  damage?: number
  attackRange?: number
  attackSpeed?: number
  attackVisual?: BuildingAttackVisual
  resourceType?: 'gold' | 'wood'
  resourceAmount?: number
  resourceCost: Record<string, number>
  capabilities: BuildingCapability[]
  spawnUnitTypes: string[]
  // Minimum town-hall tier the owning player must control to build this.
  // 0/omitted ⇒ no requirement. Mirrors the server's BuildingDef field.
  requiresTownhallTier?: number
  // Tier-chain fields (mirrors the server's BuildingDef). A tier def (e.g.
  // keep, castle) sets upgradesFrom to its previous tier and carries the cost
  // + duration to upgrade INTO it. A placed building keeps its base type plus a
  // numeric `tier` in metadata; these defs drive the upgrade action label/cost
  // and the tier display names. See getUpgradeChain.
  upgradesFrom?: string
  upgradeCost?: Record<string, number>
  upgradeSeconds?: number
  metadata: JsonObject
  color: string
  label: string
  hotkey: string
  spriteRender?: BuildingSpriteRenderDef
  selectionRing?: BuildingSelectionRingDef
  shadow?: BuildingShadowDef
}

export function getBuildingClass(def: BuildingDef | undefined | null): BuildingClass {
  return def?.class ?? 'player'
}

export let BUILDING_DEFS: BuildingDef[] = []
export let BUILDABLE_BUILDING_DEFS: BuildingDef[] = []

export let BUILDING_DEF_MAP = new Map<string, BuildingDef>()

export function initBuildingDefs(defs: BuildingDef[]): void {
  BUILDING_DEFS = defs.map((def) => ({
    ...def,
    hotkey: (def.hotkey ?? '').toLowerCase(),
  }))
  BUILDABLE_BUILDING_DEFS = BUILDING_DEFS.filter((def) => def.buildable !== false)
  BUILDING_DEF_MAP = new Map(BUILDING_DEFS.map((def) => [def.type, def]))
}

// Per-style render overrides keyed by buildingType → styleName. Populated from
// /catalog/buildings alongside the building defs. Empty for building types
// without per-sprite art.
export let BUILDING_STYLE_RENDER_MAP: Record<string, Record<string, BuildingStyleRenderDef>> = {}

export function initBuildingStyleRenders(
  styles: Record<string, Record<string, BuildingStyleRenderDef>> | null | undefined,
): void {
  BUILDING_STYLE_RENDER_MAP = styles ?? {}
}

// Returns the render override for a building type + style (e.g. recipe-shop +
// "druid-recipe-vendor"), or null when the type has no per-style config or the
// style is unset/unknown. Callers fall back to the base BuildingDef's render
// config. Style lookup is case-insensitive to match the sprite-stem keying.
export function getBuildingStyleRender(
  buildingType: string,
  style: string | null | undefined,
): BuildingStyleRenderDef | null {
  if (!style) return null
  const byStyle = BUILDING_STYLE_RENDER_MAP[buildingType]
  if (!byStyle) return null
  return byStyle[style] ?? byStyle[style.toLowerCase()] ?? null
}

// Resolves the tier chain rooted at rootType by following upgradesFrom links —
// e.g. getUpgradeChain('townhall') → [townhall, keep, castle]. Mirrors the
// server's upgradeChainFor: deterministic and capped at the catalog size so a
// malformed cycle can't loop forever. Returns [] if rootType is unknown.
export function getUpgradeChain(rootType: string): BuildingDef[] {
  const root = BUILDING_DEF_MAP.get(rootType)
  if (!root) return []
  const chain: BuildingDef[] = [root]
  while (chain.length <= BUILDING_DEFS.length) {
    const current = chain[chain.length - 1]
    const next = BUILDING_DEFS.find((def) => def.upgradesFrom === current.type)
    if (!next) break
    chain.push(next)
  }
  return chain
}

// Display name for a 1-based town-hall tier, sourced from the upgrade chain
// (tier 1 = Town Hall, 2 = Keep, 3 = Castle). Falls back to a generic label.
export function townHallTierName(tier: number): string {
  const chain = getUpgradeChain('townhall')
  return chain[tier - 1]?.label ?? `Tier ${tier}`
}

// Resolves the sprite/asset folder type for a placed building at a given 1-based
// tier. A placed building keeps its base (root) buildingType and carries a
// numeric `tier` in metadata; each higher tier renders with the art of the next
// link in its upgrade chain (townhall → keep → castle). Tier <= 1, an unknown
// root, or a chain shorter than the tier all fall back to rootType, so
// non-tiered buildings and unconfigured tiers keep their own art. Cached per
// (rootType, tier) since the catalog is static after load.
const tierSpriteTypeCache = new Map<string, string>()

export function spriteTypeForTier(rootType: string, tier: number): string {
  if (tier <= 1) return rootType
  const key = `${rootType}:${tier}`
  const cached = tierSpriteTypeCache.get(key)
  if (cached !== undefined) return cached
  const chain = getUpgradeChain(rootType)
  // Don't cache before the catalog has loaded (empty chain) — a later call once
  // BUILDING_DEFS is populated must be able to resolve the real tier art.
  if (chain.length === 0) return rootType
  const resolved = chain[tier - 1]?.type ?? rootType
  tierSpriteTypeCache.set(key, resolved)
  return resolved
}

// Footprint-derived shadow defaults (cell units), used when a building has no
// `shadow` block or leaves individual fields at zero. Mirrors the comment on
// the Go-side BuildingShadowDef.
const BUILDING_SHADOW_RADIUS_X_FROM_WIDTH = 0.42
const BUILDING_SHADOW_RADIUS_Y_FROM_RADIUS_X = 0.3
const BUILDING_SHADOW_CENTER_Y_FROM_HEIGHT = 0.85
const BUILDING_SHADOW_DEFAULT_OPACITY = 0.45

/**
 * Resolves a building's shadow config into concrete pixel draw params, or null
 * when disabled. `center` is measured from the footprint's top-left world
 * corner so the caller adds (worldX, worldY). Distance fields in the config are
 * in cell units; following the codebase convention for these building render
 * defs, a zero value means "use the footprint-derived default" (so 0 and
 * omitted behave identically).
 */
export function resolveBuildingShadow(
  shadow: BuildingShadowDef | undefined,
  widthCells: number,
  heightCells: number,
  cellSize: number,
): ResolvedBuildingShadow | null {
  if (shadow?.enabled === false) return null

  const radiusXCells = shadow?.radiusX || widthCells * BUILDING_SHADOW_RADIUS_X_FROM_WIDTH
  const radiusYCells = shadow?.radiusY || radiusXCells * BUILDING_SHADOW_RADIUS_Y_FROM_RADIUS_X
  const centerXCells = shadow?.offsetX || widthCells / 2
  const centerYCells = shadow?.offsetY || heightCells * BUILDING_SHADOW_CENTER_Y_FROM_HEIGHT
  const opacityRaw = shadow?.opacity || BUILDING_SHADOW_DEFAULT_OPACITY
  const opacity = opacityRaw < 0 ? 0 : opacityRaw > 1 ? 1 : opacityRaw

  return {
    centerX: centerXCells * cellSize,
    centerY: centerYCells * cellSize,
    radiusX: radiusXCells * cellSize,
    radiusY: radiusYCells * cellSize,
    opacity,
  }
}

export function getResolvedBuildingAttackVisual(
  def: BuildingDef | undefined | null,
): ResolvedBuildingAttackVisual {
  const inferredKind = (def?.attackRange ?? 0) > 80 ? 'projectile' : 'melee'
  const kind = def?.attackVisual?.kind ?? inferredKind
  const width = def?.width ?? 1
  const height = def?.height ?? 1

  return {
    kind,
    originX: Number((def?.attackVisual?.originX ?? width * 50).toFixed(2)),
    originY: Number((def?.attackVisual?.originY ?? height * 28).toFixed(2)),
    effectLength: Math.max(
      4,
      Math.round(def?.attackVisual?.effectLength ?? (kind === 'projectile' ? 24 : 10)),
    ),
  }
}
